File: src/VrBook.Contracts/Dtos/Pricing.cs
using VrBook.Contracts.Common;
using VrBook.Contracts.Enums;

namespace VrBook.Contracts.Dtos;

public sealed record PricingPlanDto(
    Guid Id,
    Guid PropertyId,
    decimal BaseNightlyRate,
    decimal WeekendRate,
    string Currency,
    int MinStayNights,
    int MaxStayNights,
    bool DynamicEnabled,
    IReadOnlyList<PricingRuleDto> Rules,
    IReadOnlyList<FeeDto> Fees);

public sealed record PricingRuleDto(
    Guid Id,
    PricingRuleKind Kind,
    int Priority,
    DateOnly? StartDate,
    DateOnly? EndDate,
    int? DayOfWeekMask,
    int? MinNights,
    int? MaxNights,
    int? DaysBeforeCheckin,
    PricingAdjustmentKind AdjustmentKind,
    decimal AdjustmentValue,
    bool IsEnabled);

public sealed record FeeDto(
    Guid Id,
    FeeKind Kind,
    decimal Amount,
    FeeBasis Basis,
    int? FreeThreshold,
    string Label);

public sealed record UpdatePricingPlanRequest(
    decimal BaseNightlyRate,
    decimal WeekendRate,
    string Currency,
    int MinStayNights,
    int MaxStayNights,
    bool DynamicEnabled,
    IReadOnlyList<FeeDto> Fees);

public sealed record CreatePricingRuleRequest(
    PricingRuleKind Kind,
    int Priority,
    DateOnly? StartDate,
    DateOnly? EndDate,
    int? DayOfWeekMask,
    int? MinNights,
    int? MaxNights,
    int? DaysBeforeCheckin,
    PricingAdjustmentKind AdjustmentKind,
    decimal AdjustmentValue,
    bool IsEnabled);

/// <summary>POST /properties/{id}/quotes request body.</summary>
public sealed record QuoteRequest(
    DateOnly Checkin,
    DateOnly Checkout,
    int Guests,
    bool ApplyLoyaltyDiscount);

/// <summary>Per-night and total breakdown returned by the pricing engine. See proposal §11.2.</summary>
public sealed record QuoteDto(
    Guid PropertyId,
    DateRange Range,
    int Guests,
    IReadOnlyList<NightlyLineDto> Nightly,
    IReadOnlyList<FeeLineDto> Fees,
    Money Subtotal,
    Money Discount,
    Money Taxes,
    Money Total,
    DateTimeOffset ExpiresAt);

public sealed record NightlyLineDto(DateOnly Date, Money Amount, string? RuleApplied);

public sealed record FeeLineDto(string Label, FeeKind Kind, Money Amount);

File: web/src/lib/api/pricing.ts
import { apiFetch } from './client';

// Mirror VrBook.Contracts.Common.Money / DateRange.
export interface Money {
  readonly amount: number;
  readonly currency: string;
}

export interface DateRange {
  readonly start: string;
  readonly end: string;
}

export interface NightlyLine {
  readonly date: string;
  readonly amount: Money;
  readonly ruleApplied: string | null;
}

export interface FeeLine {
  readonly label: string;
  readonly kind: string;
  readonly amount: Money;
}

export interface Quote {
  readonly propertyId: string;
  readonly range: DateRange;
  readonly guests: number;
  readonly nightly: readonly NightlyLine[];
  readonly fees: readonly FeeLine[];
  readonly subtotal: Money;
  readonly discount: Money;
  readonly taxes: Money;
  readonly total: Money;
  readonly expiresAt: string;
}

export interface QuoteRequestBody {
  readonly checkin: string;       // YYYY-MM-DD (DateOnly)
  readonly checkout: string;      // YYYY-MM-DD
  readonly guests: number;
  readonly applyLoyaltyDiscount?: boolean;
}

export const computeQuote = (propertyId: string, body: QuoteRequestBody): Promise<Quote> =>
  apiFetch<Quote>(`/api/v1/properties/${encodeURIComponent(propertyId)}/quotes`, {
    body: { applyLoyaltyDiscount: false, ...body },
    anonymous: true,
  });

// Slice 6: mirror VrBook.Contracts.Enums.PricingRuleKind / PricingAdjustmentKind.
export type PricingRuleKind =
  | 'DateRangeOverride'
  | 'LastMinute'
  | 'LengthOfStay'
  | 'DayOfWeek'
  | 'Base';

export type PricingAdjustmentKind = 'Absolute' | 'Multiplier' | 'Override';

export interface PricingRule {
  readonly id: string;
  readonly kind: PricingRuleKind;
  readonly priority: number;
  readonly startDate: string | null;
  readonly endDate: string | null;
  readonly dayOfWeekMask: number | null;
  readonly minNights: number | null;
  readonly maxNights: number | null;
  readonly daysBeforeCheckin: number | null;
  readonly adjustmentKind: PricingAdjustmentKind;
  readonly adjustmentValue: number;
  readonly isEnabled: boolean;
}

export interface CreatePricingRuleRequest {
  readonly kind: PricingRuleKind;
  readonly priority: number;
  readonly startDate: string | null;
  readonly endDate: string | null;
  readonly dayOfWeekMask: number | null;
  readonly minNights: number | null;
  readonly maxNights: number | null;
  readonly daysBeforeCheckin: number | null;
  readonly adjustmentKind: PricingAdjustmentKind;
  readonly adjustmentValue: number;
  readonly isEnabled: boolean;
}

export interface PricingPlan {
  readonly id: string;
  readonly propertyId: string;
  readonly baseNightlyRate: number;
  readonly weekendRate: number;
  readonly currency: string;
  readonly minStayNights: number;
  readonly maxStayNights: number;
  readonly dynamicEnabled: boolean;
  readonly rules: readonly PricingRule[];
  readonly fees: readonly {
    readonly id: string;
    readonly kind: string;
    readonly amount: number;
    readonly basis: string;
    readonly freeThreshold: number | null;
    readonly label: string;
  }[];
}

export const getPricingPlan = (propertyId: string): Promise<PricingPlan> =>
  apiFetch<PricingPlan>(`/api/v1/properties/${encodeURIComponent(propertyId)}/pricing`);

// --- Slice 6 rule CRUD + reorder ----------------------------------------

export const createPricingRule = (propertyId: string, body: CreatePricingRuleRequest): Promise<PricingRule> =>
  apiFetch<PricingRule>(`/api/v1/properties/${encodeURIComponent(propertyId)}/pricing/rules`, {
    method: 'POST',
    body,
  });

export const updatePricingRule = (propertyId: string, ruleId: string, body: CreatePricingRuleRequest): Promise<PricingRule> =>
  apiFetch<PricingRule>(`/api/v1/properties/${encodeURIComponent(propertyId)}/pricing/rules/${encodeURIComponent(ruleId)}`, {
    method: 'PUT',
    body,
  });

export const deletePricingRule = (propertyId: string, ruleId: string): Promise<void> =>
  apiFetch<void>(`/api/v1/properties/${encodeURIComponent(propertyId)}/pricing/rules/${encodeURIComponent(ruleId)}`, {
    method: 'DELETE',
  });

export const setPricingRuleEnabled = (propertyId: string, ruleId: string, isEnabled: boolean): Promise<PricingRule> =>
  apiFetch<PricingRule>(`/api/v1/properties/${encodeURIComponent(propertyId)}/pricing/rules/${encodeURIComponent(ruleId)}/enabled`, {
    method: 'PATCH',
    body: { isEnabled },
  });

export const reorderPricingRules = (propertyId: string, ruleIds: readonly string[]): Promise<PricingPlan> =>
  apiFetch<PricingPlan>(`/api/v1/properties/${encodeURIComponent(propertyId)}/pricing/rules/reorder`, {
    method: 'POST',
    body: { ruleIds },
  });
